daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala

```diff
@@ -240,6 +247,7 @@ object Node {
       children: ImmArray[Nid],
       exerciseResult: Option[Val],
       key: Option[KeyWithMaintainers[Val]],
+      override val byKey: Boolean,
```

Comment:
I assume there's some invariant here we should document. I would guess that when `key` is `None` then `byKey` must be false. Does the inverse also hold? If not, why do we do store key information in the node if we didn't exercise by key?

Either way, if the presence of `key` means we've exercise by key, we don't need `byKey` at all. If `key` can be present even when we've exercised by contract id, we might want to put the boolean for `byKey` into `key`.

Reply:
The invariant is `!byKey || key.isDefined`.

The reason why the `byKey` is not inside the `key` option is that this is a more drastic change, i.e. this will at least double the size of the PR. It is also a change of API that will impact ledger implementers. 

I you wish I can do it in another PR.

Reply:
I think I do now understand why we need the `byKey` field where it is. Let's keep it there and add a comment documenting the invariant between `key` and `byKey`.